mir-ci/mir_ci/tests/test_osk.py

```diff
@@ -0,0 +1,69 @@
+import itertools
+from pathlib import Path
+from typing import Collection
+
+import pytest
+from mir_ci import VARIANT
+from mir_ci.fixtures import apps
+from mir_ci.program.app import App
+from mir_ci.program.display_server import DisplayServer
+from mir_ci.program.program import ProgramError
+from mir_ci.wayland.screencopy_tracker import ScreencopyTracker
+from mir_ci.wayland.virtual_pointer import VirtualPointer
+
+ROBOT_PATH = Path(__file__).parent / "robot"
+
+
+def collect_asset(asset: Path, variant):
+    variants = (variant, *variant.parents[:-1])
+    return itertools.chain(
+        asset.glob("*"),
+        *((asset / "variants" / v).glob("*") for v in variants),
+    )
+
+
+def collect_assets(platform: str, resources: Collection[str], suite: str, variant: Path = VARIANT):
+    return {
+        p.name: p
+        for p in itertools.chain(
+            collect_asset(ROBOT_PATH / "platforms" / platform, variant),
+            *(collect_asset(ROBOT_PATH / "resources" / resource, variant) for resource in resources),
+            collect_asset(ROBOT_PATH / "suites" / suite, variant),
+        )
+    }
```

Comment:
@p-gentili have a look at how I approached the asset variants problem.

Given a variant of `noble/1920/1920x1080/ubuntu-frame`, for each of `platform`, `*resources` and `suite` it will look for the most specific assets:
```python
In [93]: variant = Path("noble/1920/1920x1080/ubuntu-frame")
    ...: ("<part>/*",) + tuple(f"<part>/variants/{p}/*" for p in
    ...:   reversed(
    ...:     (Path().joinpath(*variant.parts[i:]),
    ...:     *Path().joinpath(*variant.parts[i:]).parents[:-1])
    ...:   )
    ...: )
Out[93]: 
('<part>/*',
 '<part>/variants/1920/*',
 '<part>/variants/1920/1920x1080/*',
 '<part>/variants/1920/1920x1080/ubuntu-frame/*')
```

This feels simple and flexible enough. I've been able to use this across Ubuntu distros and different screen resolutions. As you produce your assets you can decide how deep it needs to go. In theory you can override assets from previous parts, but that is pushing it…

A more complex setup could also walk "up" from each part of the variant, but that feels overkill:
```python
In [94]: variant = Path("noble/1920/1920x1080/ubuntu-frame")
    ...: ("<part>/*",) + tuple(f"<part>/variants/{p}/*" for p in
    ...:   sorted(itertools.chain.from_iterable(
    ...:     itertools.chain(
    ...:       (
    ...:         Path().joinpath(*variant.parts[i:]),
    ...:         *Path().joinpath(*variant.parts[i:]).parents[:-1]
    ...:       )
    ...:     for i in range(0, len(variant.parts)))
    ...:   ),
    ...:   key=lambda x: len(x.parts)))
Out[94]: 
('<part>/*',
 '<part>/variants/noble/*',
 '<part>/variants/1920/*',
 '<part>/variants/1920x1080/*',
 '<part>/variants/ubuntu-frame/*',
 '<part>/variants/noble/1920/*',
 '<part>/variants/1920/1920x1080/*',
 '<part>/variants/1920x1080/ubuntu-frame/*',
 '<part>/variants/noble/1920/1920x1080/*',
 '<part>/variants/1920/1920x1080/ubuntu-frame/*',
 '<part>/variants/noble/1920/1920x1080/ubuntu-frame/*')
```